platforms/core-configuration/model-core/src/main/java/org/gradle/internal/instantiation/managed/ManagedObjectRegistry.java

```diff
@@ -0,0 +1,220 @@
+            Method interfaceMethod = null;
+            try {
+                interfaceMethod = parent.getMethod(targetMethod.getName(), targetMethod.getParameterTypes());
+            } catch (NoSuchMethodException e) {
+                // Method not found in this interface, continue checking others
+            }
+            if (interfaceMethod != null && interfaceMethod.isAnnotationPresent(annotation)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private void registerFactory(MethodHandle handle, Object instance) {
+        validateMethod(handle);
+
+        MethodType type = handle.type();
+        Class<?> returnType = type.returnType();
+        FactoryRegistration existing = factories.put(returnType, new FactoryRegistration(instance, handle));
+
+        if (existing != null) {
+            throw new IllegalArgumentException("Method " + handle + " annotated with @ManagedObjectCreator conflicts with existing factory for type " + returnType);
+        }
+    }
+
+    private static void validateMethod(MethodHandle handle) {
+        MethodType type = handle.type();
+
+        // The 0th parameter is the `this` receiver parameter.
+        int actualParameterCount = type.parameterCount() - 1;
+
+        if (actualParameterCount > 2) {
+            // We only support max 2 arg factories.
+            throw new IllegalArgumentException("Method " + handle + " annotated with @ManagedObjectCreator has too many parameters");
+        }
+
+        for (int i = 0; i < actualParameterCount; i++) {
+            Class<?> parameterType = type.parameterType(i + 1);
+            if (parameterType != Class.class) {
+                throw new IllegalArgumentException("Method " + handle + " annotated with @ManagedObjectCreator must have parameters of type Class, but has parameter of type " + parameterType);
+            }
+        }
+    }
+
+    private static class FactoryRegistration {
+        private final Object instance;
+        private final MethodHandle method;
+
+        private FactoryRegistration(Object instance, MethodHandle method) {
+            this.instance = instance;
+            this.method = method;
+        }
+    }
+
+    /**
+     * Create a managed object instance for a type with no arguments.
+     */
+    @Nullable
+    public <T> T newInstance(Class<T> type) {
+        FactoryRegistration factory = factories.get(type);
+
+        if (factory != null) {
+            try {
+                Object result = factory.method.invoke(factory.instance);
+                return type.cast(result);
+            } catch (Throwable e) {
+                throw new RuntimeException("Could not create managed object", e);
+            }
+        }
+
+        if (parent != null) {
+            return parent.newInstance(type);
+        }
+
+        return null;
+    }
+
+    /**
+     * Create a managed object instance for a type with a type argument.
```

Comment:
💭   It's unusual for a `newInstance` method to return null. I wonder if we should find a more appropriate name, since we are being lenient here.

At the very least we should add a note in the javadoc.

Reply:
I'm open to other names, but am personally happy with this one.

In any case I've added additional javadoc